### PR TITLE
fix loss of context/cause on exceptions raised inside open_websocket

### LINKED ISSUE #191: __cause__ and __context__ on user exceptions are lost when running with exception groups
This code will fully lose the `TypeError` as the `__context__` from the exception, due to https://github.com/python-trio/flake8-async/issues/298
```python
async def test_user_exception_cause(nursery) -> None:
    async def handler(request):
        await request.accept()
    server = await nursery.start(serve_websocket, handler, HOST, 0, None)
    async with open_websocket(HOST, server.port, RESOURCE, use_ssl=False):
        try:
            raise TypeError
        except TypeError:
            raise ValueError
```
The offending code for this example is
https://github.com/python-trio/trio-websocket/blob/f5fd6d77db16a7b527d670c4045fa1d53e621c35/trio_websocket/_impl.py#L205-L208
which implicitly sets `__context__` to the caught group, and the user exception inside the group is the same exception, so it simultaneously loses its old `__context__`.

The fix is to sprinkle a bunch of `copy.copy()` calls. Given that we have a `finally` it's not an alternative to raise the exception outside of the exception handler.

### PR DESCRIPTION
**Pull request by jakkdl**
fixes #191 

I'm not 100% sure that `copy_exc` is 100% robust (given that everything seems ever so fickle with copying exceptions), so it may be a good idea to wrap it in a `try: except` that gives up on copying if anything fails.

### CODE REVIEW

**Review by belm0 — APPROVED** [review]

**Review by graingert — CHANGES_REQUESTED** [review]

File: trio_websocket/_impl.py [hunk 1]
```diff
@@ -91,6 +92,16 @@ def __exit__(self, ty, value, tb):
             filtered_exception = _ignore_cancel(value)
         return filtered_exception is None
 
+def copy_exc(e: BaseException) -> BaseException:
```
[the remark this answers is not in the document]
Comment (jakkdl) [reply trio_websocket/_impl.py:95]: oops, yes. Relic of previous implementation

File: trio_websocket/_impl.py [hunk 2]
```diff
@@ -181,10 +181,27 @@ async def _close_connection(connection: WebSocketConnection) -> None:
...
+    def _raise(exc: BaseException) -> NoReturn:
+        """This helper allows re-raising an exception without __context__ being set."""
+        # cause does not need special handlng, we simply avoid using `raise .. from ..`
+        __tracebackhide__ = True
+        context = exc.__context__
+        try:
+            raise exc
+        finally:
+            exc.__context__ = context
+            del exc, context
```
Comment (graingert) [inline trio_websocket/_impl.py:193]: just remembered an issue with this:

https://github.com/python/cpython/issues/99856
https://github.com/python-attrs/attrs/pull/1081

```suggestion
        context = exc.__context__
        try:
            raise exc
        finally:
            object.__setattr__(exc, "__context__", context)
            del exc, context
```

Comment (graingert) [reply trio_websocket/_impl.py:193]: You'll probably need a test with a frozen dataclass and attrs class Exception

File: trio_websocket/_impl.py [hunk 3]
```diff
@@ -218,25 +235,40 @@ async def _close_connection(connection: WebSocketConnection) -> None:
...
+            # it completely because we can't partially suppress the group
+            if user_error.__context__ is not None and not user_error.__suppress_context__:
+                exceptions.append(user_error.__context__)
+                eg_substr = ' and the context for the user exception'
+            eg_str = (
+                "Both internal and user exceptions encountered. This group contains "
+                "the internal exception(s)" + eg_substr + "."
+            )
+            user_error.__context__ = BaseExceptionGroup(eg_str, exceptions)
+            user_error.__suppress_context__ = False
```
Comment (graingert) [inline trio_websocket/_impl.py:270]: this is broken for frozen user_errors on attrs https://github.com/python-attrs/attrs/pull/1365

File: trio_websocket/_impl.py (hunk 2 above)
Comment (jakkdl) [reply trio_websocket/_impl.py:193]: I can fix them... but `open_websocket` is decorated with `@asynccontextmanager` so it just kicks the can onto failing in the same way as https://github.com/python/cpython/issues/99856

For `attrs` compatibility I kind of feel like we can require users with frozen exception to require `trio_websocket<=0.11.1` or `attrs>24.2.0` (both to be released) - or change their exceptions. It's not a heavy workaround for us, but I still don't really think it's worth it.

File: trio_websocket/_impl.py (hunk 3 above)
Comment (jakkdl) [reply trio_websocket/_impl.py:270]: see https://github.com/python-trio/trio-websocket/pull/192#discussion_r1819118726, not sure we care to fix it.

**Comment by belm0** [comment]
you did a merge commit rather than squash & merge... 😢

**Comment by graingert** [comment]
Disable the wrong kinds of merge in the repo settings?

**Pull request merged**